### PR TITLE
Added yv2 pool for compound, cream, iron bank, and more

### PR DESCRIPTION
**Pull request by andrecronje**
Signed-off-by: Andre Cronje <andre.cronje@yearn.finance>

### What I did

Related issue: #

### CODE REVIEW

**Comment by Deepcryptodive** [comment]
Yes please =)

File: contracts/pools/yv2/StableSwapYv2.vy [hunk 1]
```diff
@@ -0,0 +1,996 @@
...
+
+
+@view
+@internal
+def _stored_rates() -> uint256[N_COINS]:
+    # exchangeRateStored * (1 + supplyRatePerBlock * (getBlockNumber - accrualBlockNumber) / 1e18)
+    result: uint256[N_COINS] = PRECISION_MUL
+    for i in range(N_COINS):
+        rate: uint256 = cyToken(self.coins[i]).exchangeRateStored()
+        rate += rate * cyToken(self.coins[i]).supplyRatePerBlock() * (block.number - cyToken(self.coins[i]).accrualBlockNumber()) / PRECISION
```
Comment (iamdefinitelyahuman) [inline contracts/pools/yv2/StableSwapYv2.vy:226]: assigning `self.coins[i]` to a memory variable will reduce the number of storage loads here, and should help the bytecode size.

File: contracts/pools/yv2/StableSwapYv2.vy [hunk 2]
```diff
@@ -0,0 +1,996 @@
...
+    return result
+
+
+@view
+@internal
+def _xp(rates: uint256[N_COINS]) -> uint256[N_COINS]:
+    result: uint256[N_COINS] = rates
+    for i in range(N_COINS):
+        result[i] = result[i] * self.balances[i] / PRECISION
+    return result
```
Comment (iamdefinitelyahuman) [inline contracts/pools/yv2/StableSwapYv2.vy:244]:
```suggestion
    result: uint256[N_COINS] = empty(uint256[N_COINS])
    for i in range(N_COINS):
        result[i] = rates[i] * self.balances[i] / PRECISION
    return result
```

We can save some gas by avoiding writing the intermediate value to memory.

File: contracts/pools/yv2/StableSwapYv2.vy [hunk 3]
```diff
@@ -0,0 +1,996 @@
...
+    token_supply: uint256 = ERC20(_lp_token).totalSupply()
+
+    # Initial invariant
+    D0: uint256 = 0
+    old_balances: uint256[N_COINS] = self.balances
+    if token_supply != 0:
+        D0 = self.get_D_mem(rates, old_balances, amp)
+
+    # Take coins from the sender
+    coin: address = ZERO_ADDRESS
```
Comment (iamdefinitelyahuman) [inline contracts/pools/yv2/StableSwapYv2.vy:360]: moving this line to line 369 (immediately after the `if amount!= 0:`) will help with the bytecode size by freeing the memory slot when the scope closes

File: contracts/pools/yv2/StableSwapYv2.vy [hunk 4]
```diff
@@ -0,0 +1,996 @@
...
+    new_balances: uint256[N_COINS] = old_balances
+    amounts: uint256[N_COINS] = empty(uint256[N_COINS])
+    for i in range(N_COINS):
+        amount: uint256 = _amounts[i]
+
+        if token_supply == 0:
+            assert amount > 0
+        if amount != 0:
```
Comment (iamdefinitelyahuman) [inline contracts/pools/yv2/StableSwapYv2.vy:368]:
```suggestion
        if amount == 0:
            assert token_supply > 0
        else:
```

File: contracts/pools/yv2/StableSwapYv2.vy [hunk 5]
```diff
@@ -0,0 +1,996 @@
...
+    dx_ = ERC20(self.coins[i]).balanceOf(self) - dx_
+    dy_: uint256 = self._exchange(i, j, dx_)
+
+    assert cyToken(self.coins[j]).redeem(dy_) == 0
+    dy: uint256 = ERC20(self.underlying_coins[j]).balanceOf(self)
+    assert dy >= min_dy, "Too few coins in result"
+
+    ERC20(self.underlying_coins[j]).transfer(msg.sender, dy)
+
+    log TokenExchangeUnderlying(msg.sender, i, dx, j, dy)
```
Comment (iamdefinitelyahuman) [inline contracts/pools/yv2/StableSwapYv2.vy:595]: repeated references to `self.coins` and `self.underlying_coins` should be handled via a memory variable to reduce storage loads

File: contracts/pools/yv2/StableSwapYv2.vy [hunk 6]
```diff
@@ -0,0 +1,996 @@
...
+        _balance: uint256 = self.balances[i]
+        value: uint256 = _balance * _amount / total_supply
+        self.balances[i] = _balance - value
+        amounts[i] = value
+
+        coin: address = self.coins[i]
+        if _use_underlying:
+            assert cyToken(coin).redeem(value) == 0
+            value = ERC20(self.underlying_coins[i]).balanceOf(self)
+            ERC20(self.underlying_coins[i]).transfer(msg.sender, value)
```
Comment (iamdefinitelyahuman) [inline contracts/pools/yv2/StableSwapYv2.vy:629]: another good spot for a memory var

File: contracts/pools/yv2/StableSwapYv2.vy [hunk 7]
```diff
@@ -0,0 +1,996 @@
...
+    assert token_amount <= _max_burn_amount, "Slippage screwed you"
+
+    CurveToken(lp_token).burnFrom(msg.sender, token_amount)  # dev: insufficient funds
+    for i in range(N_COINS):
+        amount: uint256 = amounts[i]
+        if amount != 0:
+            coin: address = self.coins[i]
+            if _use_underlying:
+                assert cyToken(coin).redeem(amount) == 0
+                ERC20(self.underlying_coins[i]).transfer(msg.sender, ERC20(self.underlying_coins[i]).balanceOf(self))
```
Comment (iamdefinitelyahuman) [inline contracts/pools/yv2/StableSwapYv2.vy:707]: and again

File: contracts/pools/yv2/StableSwapYv2.vy [hunk 8]
```diff
@@ -0,0 +1,996 @@
...
+    dy, dy_fee = self._calc_withdraw_one_coin(_token_amount, i, False)
+    amount: uint256 = dy
+
+    self.balances[i] -= (dy + dy_fee * self.admin_fee / FEE_DENOMINATOR)
+    CurveToken(self.lp_token).burnFrom(msg.sender, _token_amount)  # dev: insufficient funds
+    coin: address = self.coins[i]
+    if _use_underlying:
+        assert cyToken(coin).redeem(dy) == 0
+        amount = ERC20(self.underlying_coins[i]).balanceOf(self)
+        ERC20(self.underlying_coins[i]).transfer(msg.sender, amount)
```
Comment (iamdefinitelyahuman) [inline contracts/pools/yv2/StableSwapYv2.vy:845]: last one

**Review by iamdefinitelyahuman — COMMENTED** [review]
Looking good so far, mostly just some gas golfing to do.

File: contracts/pools/yv2/StableSwapYv2.vy [hunk 9]
```diff
@@ -0,0 +1,1009 @@
...
+    dx_: uint256 = ERC20(coin).balanceOf(self)
+    assert cyToken(coin).mint(dx) == 0
+    dx_ = ERC20(coin).balanceOf(self) - dx_
+    dy_: uint256 = self._exchange(i, j, dx_)
+
+    assert cyToken(self.coins[j]).redeem(dy_) == 0
+
+    underlying: address = self.underlying_coins[j]
+
+    dy: uint256 = ERC20(underlying).balanceOf(self)
```
Comment (michwill) [inline contracts/pools/yv2/StableSwapYv2.vy:600]: Just a very bizare idea (though it doesn't change too much). If we add `- 1` here, storage becomes cheaper, and we save 15k gas

**Review by michwill — APPROVED** [review]
Very nice and solid PR

**Pull request merged**